core/src/main/java/org/apache/druid/java/util/common/guava/ParallelMergeCombiningSequence.java

```diff
@@ -0,0 +1,1077 @@
+   * for each partition a {@link PrepareMergeCombineInputsAction} will be executed to to wait for each of the yielders to
+   * yield {@link ResultBatch}. After the cursors all have an initial set of results, the
+   * {@link PrepareMergeCombineInputsAction} will execute a {@link MergeCombineAction}
+   * to perform the actual work of merging sequences and combining results. The merged and combined output of each
+   * partition will itself be put into {@link ResultBatch} and pushed to a {@link BlockingQueue} with a
+   * {@link ForkJoinPool} {@link QueuePusher}.
+   *
+   * The second layer will execute a single {@link PrepareMergeCombineInputsAction} to wait for the {@link ResultBatch}
+   * from each partition to be available in their 'output' {@link BlockingQueue} which each is wrapped in
+   * {@link BlockingQueueuBatchedResultsCursor}. Like the first layer, after the {@link PrepareMergeCombineInputsAction}
+   * is complete and some {@link ResultBatch} are ready to merge from each partition, it will execute a
+   * {@link MergeCombineAction} do a final merge combine of all the parallel computed results, again pushing
+   * {@link ResultBatch} into a {@link BlockingQueue} with a {@link QueuePusher}.
+   */
+  private static class MergeCombinePartitioningAction<T> extends RecursiveAction
+  {
+    private final List<Sequence<T>> sequences;
+    private final Ordering<T> orderingFn;
+    private final BinaryOperator<T> combineFn;
+    private final BlockingQueue<ResultBatch<T>> out;
+    private final int queueSize;
+    private final int parallelism;
+    private final int yieldAfter;
+    private final int batchSize;
+    private final boolean hasTimeout;
+    private final long timeoutAt;
+    private final CancellationGizmo cancellationGizmo;
+
+    private MergeCombinePartitioningAction(
+        List<Sequence<T>> sequences,
+        Ordering<T> orderingFn,
+        BinaryOperator<T> combineFn,
+        BlockingQueue<ResultBatch<T>> out,
+        int queueSize,
+        int parallelism,
+        int yieldAfter,
+        int batchSize,
+        boolean hasTimeout,
+        long timeoutAt,
+        CancellationGizmo cancellationGizmo
+    )
+    {
+      this.sequences = sequences;
+      this.combineFn = combineFn;
+      this.orderingFn = orderingFn;
+      this.out = out;
+      this.queueSize = queueSize;
+      this.parallelism = parallelism;
+      this.yieldAfter = yieldAfter;
+      this.batchSize = batchSize;
+      this.hasTimeout = hasTimeout;
+      this.timeoutAt = timeoutAt;
+      this.cancellationGizmo = cancellationGizmo;
+    }
+
+    @Override
+    protected void compute()
+    {
+      try {
+        final int parallelTaskCount = computeNumTasks();
+
+        // if we have a small number of sequences to merge, or computed paralellism is too low, do not run in parallel,
+        // just serially perform the merge-combine with a single task
+        if (sequences.size() < 4 || parallelTaskCount < 2) {
```

Comment:
started skimming through the code here, so my understanding might be very superficial right now.

but, it seems sequences.size() check can be done in `CachingClusteredClient` and not use the parallel flow at all and let jetty thread do the work ... doing the check here means work is still done async for no reason ... no ?

Reply:
well, it is not for no reason, it is to have (nearly) all work being done on the same thread pool instead of having 2 competing thread pools of different sizes doing the work

Reply:
Ideally I would make scan queries run on this too, so that way all merges are done on the same pool, similar in spirit to the processing pools in the historicals i guess.

Reply:
got it, makes sense.

Reply:
Also, thanks for having a look!

Reply:
BTW, I think the check should be `sequences.size() <= 4` because following code

```
      final int computedOptimalParallelism = Math.min(
          (int) Math.floor((double) sequences.size() / 2.0),
          computedParallelism - 1
      );
```
would always return a value less/equal to 1 for sequences.size() == 3 .

Reply:
Ah, I think that does make the check against `sequences.size()` done after this redundant:

```
        final int parallelTaskCount = computeNumTasks();

        // if we have a small number of sequences to merge, or computed paralellism is too low, do not run in parallel,
        // just serially perform the merge-combine with a single task
        if (sequences.size() < 4 || parallelTaskCount < 2) {
...
```
and can just use parallelTaskCount.